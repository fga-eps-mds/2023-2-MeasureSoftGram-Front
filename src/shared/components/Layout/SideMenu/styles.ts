import styled from 'styled-components';

import { Button } from '@mui/material';

export const Wrapper = styled.div`
  height: 100vh;
  width: inherit;

  display: flex;
  flex-direction: column;
  align-items: flex-start;

  padding: 30px 12px;
  left: 0;

  background-color: white;

  position: fixed;
`;

export const Ul = styled.ul`
  width: 100%;
  display: flex;
  flex-direction: column;
  margin: 0;
  padding: 30px 0 0 0;
`;

export const Li = styled.li`
  list-style-type: none;
  padding-bottom: 2px;

  &:not(:first-child) {
    margin-top: 8px;
  }
`;

export const Logo = styled.img`
  cursor: pointer;
`;

export const NavButton = styled(Button)`
  justify-content: flex-start;

  :disabled {
    background-color: #f2f2f2;
  }
`;
