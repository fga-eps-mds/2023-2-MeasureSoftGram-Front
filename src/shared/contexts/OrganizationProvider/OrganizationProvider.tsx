import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback } from 'react';

import { Organization } from '@customTypes/organization';
import { organizationQuery } from '@services/organization';
import { useRequest } from '@hooks/useRequest';
import { useRouter } from 'next/router';

interface Props {
  children: ReactNode;
}

interface IOrganizationContext {
  currentOrganization: Organization | undefined;
  setCurrentOrganization: (organization: Organization) => void;
  organizationList: Organization[];
}

const OrganizationContext = createContext<IOrganizationContext | undefined>(undefined);

export function OrganizationProvider({ children }: Props) {
  const router = useRouter();

  const [currentOrganization, setCurrentOrganization] = useState<Organization | undefined>();

  const { data } = useRequest<{ results: [Organization] }>(organizationQuery.getAllOrganization());

  const organizationList = useCallback(() => data?.results ?? [], [data])();

  const value = useMemo(() => {
    const regex = /\d+/g;
    const queryProduct = router.query?.product as string;
    const organizationIndex = regex.exec(queryProduct);
    if (!currentOrganization && organizationList.length > 0 && queryProduct && organizationIndex) {
      setCurrentOrganization(organizationList[parseInt(organizationIndex[0], 10) - 1]);
    }
    return { currentOrganization, setCurrentOrganization, organizationList };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentOrganization, organizationList]);

  return <OrganizationContext.Provider value={value}>{children}</OrganizationContext.Provider>;
}

export function useOrganizationContext() {
  const context = useContext(OrganizationContext);

  if (context === undefined) {
    throw new Error('OrganizationContext must be used within a OrganizationContext');
  }

  return context;
}
