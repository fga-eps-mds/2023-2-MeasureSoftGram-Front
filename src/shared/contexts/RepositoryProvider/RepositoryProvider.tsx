import React, { createContext, useState, useContext, ReactNode, useMemo } from 'react';

import { Repository, Historical } from '@customTypes/repository';

interface Props {
  children: ReactNode;
}

interface IRepositoryContext {
  currentRepository?: Repository;
  setCurrentRepository: (repository: Repository) => void;
  repositoryList?: Repository[];
  setRepositoryList: (repository: Repository) => void;
  characteristics: string[];
  setCharacteristics: (characteristics: string[]) => void;
  subCharacteristics: string[];
  setSubCharacteristics: (subCharacteristics: string[]) => void;
  measures: string[];
  setMeasures: (measures: string[]) => void;
  metrics: string[];
  setMetrics: (metrics: string[]) => void;
  historicalTSQMI: Historical;
  setHistoricalTSQMI: (historical: Historical) => void;
}

const RepositoryContext = createContext<IRepositoryContext | undefined>(undefined);

export function RepositoryProvider({ children }: Props) {
  const [currentRepository, setCurrentRepository] = useState<Repository | undefined>();
  const [repositoryList, setRepositoryList] = useState<Repository[]>();

  const [characteristics, setCharacteristics] = useState<string[]>([]);
  const [subCharacteristics, setSubCharacteristics] = useState<string[]>([]);
  const [measures, setMeasures] = useState<string[]>([]);
  const [metrics, setMetrics] = useState<string[]>([]);
  const [historicalTSQMI, setHistoricalTSQMI] = useState<Historical>();

  const value = useMemo(
    () => ({
      currentRepository,
      setCurrentRepository,
      repositoryList,
      setRepositoryList,
      characteristics,
      setCharacteristics,
      subCharacteristics,
      setSubCharacteristics,
      measures,
      setMeasures,
      metrics,
      setMetrics,
      historicalTSQMI,
      setHistoricalTSQMI
    }),
    [currentRepository, repositoryList, characteristics, subCharacteristics, measures, metrics, historicalTSQMI]
  );

  return <RepositoryContext.Provider value={value}>{children}</RepositoryContext.Provider>;
}

export function useRepositoryContext() {
  const context = useContext(RepositoryContext);

  if (context === undefined) {
    throw new Error('useRepositoryContext must be used within a RepositoryContext');
  }

  return context;
}
