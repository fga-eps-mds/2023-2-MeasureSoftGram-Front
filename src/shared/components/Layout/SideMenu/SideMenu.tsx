import React, { useEffect, useState } from 'react';
import Link from 'next/link';
import { useRouter } from 'next/router';

import { Typography } from '@mui/material';

import OrganizationButton from './OrganizationSelector';
import { HEADER } from './consts';

import * as Styles from './styles';

const { MENU_OPTIONS, IMAGE_SOURCE } = HEADER.VALUES;

function Header() {
  const router = useRouter();
  const [currentPath, setCurrentPath] = useState('');

  useEffect(() => {
    const { asPath } = router;

    const pathnameArray = asPath.split('/');
    setCurrentPath(pathnameArray[1]);
  }, [router.pathname]);

  const handleClick = (path: string) => {
    router.push(path);
  };

  const isDisabled = (permissions: string[]): boolean => permissions.includes(currentPath);

  return (
    <Styles.Wrapper>
      <Link href="/">
        <Styles.Logo src={IMAGE_SOURCE} height={30} />
      </Link>

      <OrganizationButton />

      <Styles.Ul>
        {MENU_OPTIONS.map((option) => (
          <Styles.Li key={option.path}>
            <Styles.NavButton
              fullWidth
              disabled={isDisabled(option.permission)}
              onClick={() => handleClick(option.path)}
            >
              <option.icon />
              <Typography variant="button" pl="8px">
                {option.name}
              </Typography>
            </Styles.NavButton>
          </Styles.Li>
        ))}
      </Styles.Ul>
    </Styles.Wrapper>
  );
}

export default Header;
