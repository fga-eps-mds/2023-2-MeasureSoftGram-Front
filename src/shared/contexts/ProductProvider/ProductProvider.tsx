import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback, useEffect } from 'react';

import { Product } from '@customTypes/product';
import { productQuery } from '@services/product';
import { useOrganizationContext } from '@contexts/OrganizationProvider';

interface Props {
  children: ReactNode;
}

export interface IProductContext {
  currentProduct?: Product;
  setCurrentProduct: (product: Product) => void;
  productsList?: Product[];
  updateProductList: (products: Product[]) => void;
}

export const ProductContext = createContext<IProductContext | undefined>(undefined);

export function ProductProvider({ children }: Props) {
  const [currentProduct, setCurrentProduct] = useState<Product | undefined>();
  const [productsList, setProductsList] = useState<Product[]>([]);

  const { currentOrganization } = useOrganizationContext();

  const updateProductList = useCallback((products: Product[]) => {
    setProductsList(products);
  }, []);

  const loadAllProducts = async () => {
    try {
      if (!currentOrganization) {
        updateProductList([]);
        return;
      }

      const result = await productQuery.getAllProducts(currentOrganization.id);

      updateProductList(result.data.results);
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (currentOrganization) {
      loadAllProducts();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentOrganization]);

  const value = useMemo(
    () => ({
      currentProduct,
      setCurrentProduct,
      productsList,
      updateProductList
    }),
    [currentProduct, productsList, updateProductList]
  );

  return <ProductContext.Provider value={value}>{children}</ProductContext.Provider>;
}

export function useProductContext() {
  const context = useContext(ProductContext);

  if (!context) {
    throw new Error('useProductContext must be used within a ProductContext');
  }

  return context;
}