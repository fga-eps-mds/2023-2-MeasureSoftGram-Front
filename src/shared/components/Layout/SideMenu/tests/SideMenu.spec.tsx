import { findAllByText, findByText, render, } from '@testing-library/react';
import React, { ReactNode } from 'react';
import { RepositoryProvider } from '@contexts/RepositoryProvider';
import { OrganizationProvider } from '@contexts/OrganizationProvider';
import * as ProductProvApi from '@contexts/ProductProvider';
import { SideMenuProvider } from '@contexts/SidebarProvider/SideMenuProvider';
import SideMenu, { SideMenuItemType } from '../SideMenu';
import { Product } from '@customTypes/product';
import { IProductContext } from '@contexts/ProductProvider';

interface Props {
  children: ReactNode;
}

const AllTheProviders = ({ children }: Props) => (
  <OrganizationProvider>
    <ProductProvApi.ProductProvider>
      <RepositoryProvider>
        <SideMenuProvider>{children}</SideMenuProvider>
      </RepositoryProvider>
    </ProductProvApi.ProductProvider>
  </OrganizationProvider>
);

jest.mock('@contexts/ProductProvider', () => ({
    __esModule: true,
    ...jest.requireActual('@contexts/ProductProvider')
}));

jest.mock('../SideMenuWrapper', () => {
  return jest.fn().mockImplementation(() => {
    const products: Product[] = [
      {id: 'prodtest', name: 'prodname', description: 'proddesc', github_url: 'https://test.github.com', created_at: '2022-01-01', updated_at: '2022-01-02'},
      {id: 'prodtest2', name: 'prodname2', description: 'proddesc2', github_url: 'https://test.github.com/proj2', created_at: '2022-02-01', updated_at: '2022-02-02'}
    ]; 

    const productsContext: IProductContext = {
          productsList: products,
          currentProduct: undefined,
          setCurrentProduct: jest.fn(),
          updateProductList: jest.fn()
        }

    const menuItems: SideMenuItemType[] = [
      {startIcon:<p> icon </p>, text: 'test', tooltip: 'testTool', path: 'test/path', disable: false, selected: true},
      {startIcon: <p> cow </p>, text: 'different', tooltip: 'diffTool', path: 'diff/dest', disable: false, selected: true},
    ];

    let count = 0;
    return <div data-testid="mocked-wrapper">
      {menuItems.map((item) => 
        productsContext.currentProduct && item.disable ? <p key={count++}>{item.text}</p> : <p key={(count++) + 10}>Null element</p>)}
    </div>
  });
});

jest.spyOn(ProductProvApi,'useProductContext').mockReturnValue(
    { 
      currentProduct: { 
        id: 'prodtest', 
        name: 'prodname', 
        description: 'proddesc',
        github_url: 'https://test.github.com',
        created_at: '2022-01-01', 
        updated_at: '2022-01-02'
      } as Product
    } as IProductContext
)
describe('SideMenuItem', () => {
  it('should render the SideMenuItem component', () => {
    const { container } = render(<SideMenu/>, {
      wrapper: AllTheProviders
    });

    expect(container).toMatchSnapshot();
  });

  it('should render the SideMenuItem component with endIcon', () => {
    const { container } = render(
      <SideMenu/>,
      { wrapper: AllTheProviders }
    );

    expect(container).toMatchSnapshot();
  });

  it('should not render without current product', async() => {

    const { container } = render(
      <SideMenu />,
      { wrapper: AllTheProviders }
    );

    const elem = await findAllByText(container, 'Null element');
    expect(elem).not.toBeNull();
    expect(ProductProvApi.useProductContext).toHaveBeenCalled();
  });
});