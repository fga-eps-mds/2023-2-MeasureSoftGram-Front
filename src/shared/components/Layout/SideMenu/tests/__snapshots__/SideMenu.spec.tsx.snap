// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenuItem should render the SideMenuItem component 1`] = `
<div>
  <div
    data-testid="mocked-wrapper"
  >
    <p>
      Null element
    </p>
    <p>
      Null element
    </p>
  </div>
</div>
`;

exports[`SideMenuItem should render the SideMenuItem component with endIcon 1`] = `
<div>
  <div
    data-testid="mocked-wrapper"
  >
    <p>
      Null element
    </p>
    <p>
      Null element
    </p>
  </div>
</div>
`;
